exatomic/adf/tape21.py

```diff
@@ -0,0 +1,362 @@
+        start = fitem[idx] + 2
+        col = min(len(self[start].split()), dim)
+        stop = np.ceil(start + dim / col).astype(np.int64)
+        return self.pandas_dataframe(start, stop, col).stack().values
+
+    def parse_frequency(self, cart=True):
+        '''
+        ADF frequency parser.
+
+        Note:
+            This will toss a warning if it cannot find the mass-weighted normal modes
+            which must be used to generate the displaced structures for vibrational
+            averaging. Also, it will be unable to calculate the reduced masses as it will
+            have normalized cartesian coordinates where it expects normalized
+            mass-weighted cartesian normal modes.
+
+        Args:
+            cart (:obj:`bool`, optional): Parse the normalized cartesian coordinates or
+                                          the mass-weighted normal modes. Defaults to
+                                          :code:`True`.
+        '''
+        # search flags
+        _renorm = "NormalModes_RAW"
+        _recartnorm = "Normalmodes"
+        _refreq = "Frequencies"
+        _refreqexc = r"\bFrequencies\b"
+        _rekey = r"\bFreq\b"
+        found = self.find(_refreq, _renorm, _recartnorm, keys_only=True)
+        key = self.regex(_rekey, _refreqexc, keys_only=True)
+        # need to do this to ensure that we only match the data in the Freq block
+        found_freq = []
+        for k in key[_rekey]:
+            for f in found[_refreq]:
+                if f-1 == k:
+                    found_freq.append(f)
+                    break
+        if not found_freq:
+            return
+        found[_refreq] = found_freq
+        if not found[_refreq]:
+            return
+        if not hasattr(self, 'atom'):
+            self.parse_atom()
+        # get the number of atoms
+        nat = self.atom.last_frame.shape[0]
+        # get the frequencies
+        freq = self._dfme(found[_refreq], nat*3)
+        # find where the frequencies are zero
+        # these should be the ones that ADF determines to be translations and rotations
+        # TODO: need a test case with one imaginary frequency
+        low = freq != 0
+        nlow = low.shape[0]
+        # get only the ones that are non-zero
+        freq = freq[low]
+        nmodes = freq.shape[0]
+        freq = np.repeat(freq, nat)
+        if found[_renorm] and not cart:
+            # get the mass-weighted normal modes
+            ndisps = int(self[found[_renorm][0]+1].split()[0])
+            normalmodes = self._dfme(np.array(found[_renorm]), ndisps, idx=0)
+            calc_rmass = True
+        elif found[_recartnorm] and cart:
+            # get the non-mass-weighted normal modes and toss warning
+            ndisps = int(self[found[_recartnorm][0]+1].split()[0])
+            normalmodes = self._dfme(np.array(found[_recartnorm]), ndisps, idx=0)
+            calc_rmass = True
+        else:
+            raise Exception("Something went wrong")
```

Comment:
I would suggest dumping the values of `found`, `_renorm`, `_recartnorm` , etc. to make the exception message more valuable (also `ValueError`?).

Reply:
I agree. I made a change in the latest commit to use the `MissingSection` error as that is the only time it should ever get to that else unless there's a bug somewhere.